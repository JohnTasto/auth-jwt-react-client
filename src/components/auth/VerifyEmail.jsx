import React, { Component } from 'react'
import { connect } from 'react-redux'
import { actions as authActions } from '../../ducks/auth'


export class VerifyEmail extends Component {
  constructor(props) {
    super(props)
    this.state = {}
  }

  componentWillMount() {
    const { verifyEmail, match, authenticated, redirectLocation, setRedirect } = this.props
    if (!redirectLocation) {
      setRedirect({ pathname: '/' })
    }
    if (authenticated) {
      this.setState({ preauthed: true })
    } else {
      verifyEmail(match.params.token)
        .catch(error => this.setState({ error: error.errors._error }))  // eslint-disable-line no-underscore-dangle, max-len
    }
  }

  componentWillUnmount() {
    clearTimeout(this.timeoutId)
  }

  redirect = () => {
    const { history, redirectLocation, setRedirect } = this.props
    history.push(redirectLocation)
    setRedirect(undefined)
  }

  render() {
    const { authenticated } = this.props
    const { preauthed, error } = this.state

    if (preauthed) {
      return (
        <div>
          <p>Your email has already been verified.</p>
        </div>
      )
    }

    if (authenticated) {
      this.timeoutId = setTimeout(this.redirect, 5000)
      return (
        <div>
          <p>Success! Redirecting in five....</p>
        </div>
      )
    }

    if (error) {
      return (
        <div className="alert alert-danger">
          <strong>Error:&nbsp;</strong>
          <span>{error}</span>
        </div>
      )
    }

    return (
      <div>
        <p>Verifying your email....</p>
      </div>
    )
  }
}

export default connect(
  state => ({
    authenticated: state.auth.authenticated,
    redirectLocation: state.auth.redirectLocation,
  }),
  authActions,
  (...props) =>
    Object.assign({}, ...props),
)(VerifyEmail)
