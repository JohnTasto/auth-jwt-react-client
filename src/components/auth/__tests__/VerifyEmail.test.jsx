/* eslint-env jest */
/* eslint-disable comma-dangle */

import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

import { VerifyEmail as VerifyEmailComp } from '../VerifyEmail'


describe('<VerifyEmail />', () => {

  test('Renders message indicating email verification is in progress', () => {
    const props = {
      setRedirect: jest.fn(),
      verifyEmail: jest.fn(() => Promise.resolve()),
      match: { params: { token: '42' } }
    }
    const wrapper = shallow(<VerifyEmailComp {...props} />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  test('Calls action auth.verifyEmail with token on mount', () => {
    const props = {
      setRedirect: jest.fn(),
      verifyEmail: jest.fn(() => Promise.resolve()),
      match: { params: { token: '42' } }
    }
    shallow(<VerifyEmailComp {...props} />)
    expect(props.verifyEmail).toHaveBeenCalledWith(props.match.params.token)
  })

  test('Renders redirect message if authentication state changes to true', () => {
    const props = {
      setRedirect: jest.fn(),
      verifyEmail: jest.fn(() => Promise.resolve()),
      match: { params: { token: '42' } }
    }
    const wrapper = shallow(<VerifyEmailComp {...props} />)
    wrapper.setProps({ authenticated: true })
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  test('Renders message indicating email has already been verified if authenticated', () => {
    const wrapper = shallow(
      <VerifyEmailComp
        authenticated
        setRedirect={jest.fn()}
      />
    )
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  test('Calls action auth.setRedirect if no location has been set', () => {
    const props = {
      authenticated: true,
      setRedirect: jest.fn(),
    }
    shallow(<VerifyEmailComp {...props} />)
    expect(props.setRedirect).toHaveBeenCalled()
  })

  // it would be better to have verifyEmail: jest.fn(() => Promise.reject(...)
  // but I couldn't figure out how to get enzyme to wait for the promise
  test('Renders error on error', () => {
    const props = {
      setRedirect: jest.fn(),
      verifyEmail: jest.fn(() => Promise.resolve()),
      match: { params: { token: '42' } }
    }
    const wrapper = shallow(<VerifyEmailComp {...props} />)
    wrapper.setState({ error: 'ERROR!' })
    expect(toJson(wrapper)).toMatchSnapshot()
  })
})
