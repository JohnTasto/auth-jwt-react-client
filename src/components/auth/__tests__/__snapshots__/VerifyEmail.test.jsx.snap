// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VerifyEmail /> Renders error on error 1`] = `
<div
  className="alert alert-danger"
>
  <strong>
    Error:Â 
  </strong>
  <span>
    ERROR!
  </span>
</div>
`;

exports[`<VerifyEmail /> Renders message indicating email has already been verified if authenticated 1`] = `
<div>
  <p>
    Your email has already been verified.
  </p>
</div>
`;

exports[`<VerifyEmail /> Renders message indicating email verification is in progress 1`] = `
<div>
  <p>
    Verifying your email....
  </p>
</div>
`;

exports[`<VerifyEmail /> Renders redirect message if authentication state changes to true 1`] = `
<div>
  <p>
    Success! Redirecting in five....
  </p>
</div>
`;
